[Poomklao] iP

parnikkapore opened the pull request:
*Note: This PR is opened as part of course requirements.*

# y

> [the PR] provides a convenient way for us to access the current state of all your iP code from one location.

1. Training
1. Workload padding (not)
1. Because It's Fun!™ 😁

# Try it out

Builds are provided for selected revisions in the [Releases section](https://github.com/parnikkapore/ip/releases).

# Regression testing

To run the regression tests, run the following from the clone's root directory:

```bash
cd text-ui-test
./runtest.sh
```

Support for regression testing under Windows is currently broken.

# Todo

* [ ] Fix regression testing on Windows (`runtest.bat` needs an update)
* [ ] Make a GUI

# Misc notes

* 90s mascot energy (well that's what Duke was, isn't he)
* Plugin system was originally for completely different and decoupled functionalities

Commit:
Add Gradle support

parnikkapore committed:
Add recommended .gitignore entries

parnikkapore committed:
Add IOFormat

parnikkapore committed:
Implement echo back and exit commands

parnikkapore committed:
use box drawing characters for dividing line

parnikkapore committed:
make em a todo list

parnikkapore committed:
Make command functions take rest strings
also handled some partial inputs!

parnikkapore committed:
migrate to Task and List lines

parnikkapore committed:
Revert command API back to scanner

parnikkapore committed:
Add task mark and unmark

parnikkapore committed:
Migrate to Todo subclass

parnikkapore committed:
Use 1-indexed task numbers

parnikkapore committed:
Add deadline tasks

parnikkapore committed:
Add event tasks

parnikkapore committed:
Update test case

parnikkapore committed:
Handle missing todo name

parnikkapore committed:
Add MessagefulException

parnikkapore committed:
Grammar

parnikkapore committed:
Handle OOB task numbers for (un)marking

parnikkapore committed:
Add delete command

parnikkapore committed:
Unify todo ID reader
honestly I think this makes it unnecessarily longer but oh well

parnikkapore committed:
Add JavaDoc to public methods and classes

parnikkapore committed:
Add FallibleFunction

There is code duplication in Todos.cmdAdd* as I didn't think you can use functional features of Java with functions that throw exceptions. Turns out, you can, you just need to define your own FunctionalInterface!

Let's combine cmdAdd* into one function and pass in the factory functions instead.

parnikkapore committed:
Move MessagefulException handling to Duke.java

parnikkapore committed:
Change package root name to "duke"

See https://se-education.org/guides/conventions/java/intermediate.html#naming

parnikkapore committed:
Add basic skeleton of saving tasks

parnikkapore committed:
Refactor out flatpacking constants

parnikkapore committed:
Add task loading prototype

parnikkapore committed:
Add task saving and loading to other types

parnikkapore committed:
Parse deadline dates and times

parnikkapore committed:
Merge branch 'branch-Level-7'

parnikkapore committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/duke/modules/todos/Deadline.java

parnikkapore committed:
Clear savefile before test

parnikkapore committed:
Refactor out Storage class

parnikkapore committed:
Rename IOFormat to Ui and add error printers

IOFormat is Ui after all ;)

parnikkapore committed:
Refactor out "parser"

The parser is in fact the core logic, but we'll keep that between the two of us

parnikkapore committed:
Refactor out TaskList

parnikkapore committed:
Move entrypoint into package

parnikkapore committed:
Mass optimize imports

parnikkapore committed:
Lint

parnikkapore committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

parnikkapore committed:
Add Gradle support

parnikkapore committed:
Fix running end-to-end tests

parnikkapore committed:
Add unit tests for Todo

parnikkapore committed:
Add unit tests for MessagefulException

parnikkapore committed:
Add JavaDoc for all public methods

parnikkapore committed:
Lint

parnikkapore committed:
Add find tasks feature

parnikkapore committed:
Update regression tests for find command

parnikkapore committed:
Merge branch 'branch-A-JavaDoc'

parnikkapore committed:
Merge branch 'branch-A-CodingStandard'

parnikkapore committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/modules/todos/TaskList.java

parnikkapore committed:
Add missing JavaDoc

parnikkapore committed:
Rename MessagefulException.message to hint

Let's bring the class in compliance with the coding standard.

parnikkapore committed:
Encapsulate execute() result in class

This also means the boolean value that indicates if the app should exit is also named according to the coding standard.

parnikkapore committed:
Add JavaFX subsystem

parnikkapore committed:
Implement chat sending stub

parnikkapore committed:
Refactor out ExecuteResult

parnikkapore committed:
Separate text printing and formatting

parnikkapore committed:
Refactor to return instead of print execution result

parnikkapore committed:
Integrate returned result into GUI

parnikkapore committed:
Downgrade JavaFX version in FXML

Fixes a few warnings

parnikkapore committed:
Add hello message to GUI version

parnikkapore committed:
Add new task list message to GUI version

alongside lots of init refactoring to support returning messages during init period

parnikkapore committed:
Fix an unintentional change in the messages' formatting

parnikkapore committed:
Split the GUI reply into 5 line groups

parnikkapore committed:
Add window resizing constraints

parnikkapore committed:
Delete dead class

parnikkapore committed:
Merge branch 'branch-Level-10'

jasonqiu212 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,42 @@
+
+        printLines(messages);
+
+        boolean fExit = false;
```
does `fExit` mean forceExit? Perhaps a more intuitive boolean variable name here?

parnikkapore replied:
I think it stood for `flag_Exit`, but yeah, this should be renamed `willExit`.

jasonqiu212 commented on src/main/java/duke/ExecuteResult.java:
```diff
@@ -0,0 +1,37 @@
+    /**
+     * Lines of text to be sent as a reply from this command
+     */
+    private final List<String> reply;
```
would it be better to name this in plural form, since it's a collection of objects?

parnikkapore replied:
You do have a point, as it's not treated as a single object all the way down.

jasonqiu212 commented on src/main/java/duke/FallibleFunction.java:
```diff
@@ -0,0 +1,21 @@
+ * @param <R> the type of the input to the function.
+ */
+@FunctionalInterface
+public interface FallibleFunction<T, R> {
```
I really like this implementation of functional programming!

jasonqiu212 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,76 @@
+        if (DEV_MODE) {
+            return sayAsError(e.toString());
+        } else {
+            return sayAsError(e.getHint());
```
i like that you have a dev_mode

jasonqiu212 commented on src/main/java/duke/modules/todos/Event.java:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+
+    public static final String typeCode = "E";
```
i think the `typeCode` should be in all caps separated by underscores, since it's a constant. A comment here would be good to explain the purpose of this constant

parnikkapore replied:
You have a point. I agree about needing the comment.

jasonqiu212 commented on src/main/java/duke/gui/MainWindow.java:
```diff
@@ -0,0 +1,79 @@
+    @FXML
+    private Button sendButton;
+
+    private Image userImage = new Image(this.getClass().getResourceAsStream("/media/Damith.png"));
```
nice hahaha

jasonqiu212 reviewed (commented):
hello! your code looks great to me! I especially like the usage of functional programming and dev_mode! just some minor coding standard issues here and there 👍

parnikkapore committed:
Make printLines not print anything when sent empty list

parnikkapore committed:
Fix bubble sizing and remove bubble splitting

parnikkapore committed:
Add bye command handling in GUI

parnikkapore committed:
Fix misleading plugin loading code

parnikkapore committed:
Update main class name

parnikkapore committed:
Fix GUI app crashing upon sending command

parnikkapore committed:
Fix GUI app crashing upon sending command

parnikkapore committed:
Fix GUI app crashing upon sending command

parnikkapore committed:
Add a couple of assertions

parnikkapore committed:
Refactor out Task.fromFlatpack

Storage::loadList handles part of the correct logic for loading a
packed task list, which violates SLAP (GLAP?) and causes the function
to exceed the recommended length.

Let's refactor out the code that generates the Task of the correct type
into a static method of Task where it belongs.

Readers: the term "flatpack" refers to a losslessly restorable
serialization of a data structure into a flat list, as well as the
action of performing such a serialization.

parnikkapore committed:
Merge pull request #1 from parnikkapore/branch-A-Assertions

A-Assertions

parnikkapore committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/modules/todos/Storage.java

parnikkapore committed:
Merge pull request #2 from parnikkapore/branch-A-CodeQuality

A-CodeQuality

parnikkapore committed:
Change Event date parsing to JChronic

The date parser for Event is a very basic one that can only accept ISO
8601 style strings. This is incredibly impractical for any real user.

Let's use the JChronic library to parse the date for the event instead.
This will provide recognition for all kinds of natural date inputs.

Note that this will remove support for tracking event times.
Unfortunately, I can't find a natural date and time parser that does
not use AI magic.

parnikkapore committed:
Handle unparseable dates

parnikkapore committed:
Format event dates naturally

parnikkapore committed:
Update regression test

parnikkapore committed:
Merge branch 'jchronic'

parnikkapore committed:
Add Gradle command for running with CLI

parnikkapore committed:
Handle EOF

Sending EOF (such as by pressing C-D or forgetting to bind the input
while using Gradle) crashes the app.

Let's check for that condition and exit the app immediately and cleanly
instead.

parnikkapore committed:
Migrate regression test to Gradle

The old test code cannot run because the Maven-based libraries are not
in the search path.

Let's fix that by using Gradle to compile and execute the CLI version.

parnikkapore committed:
Rename variables according to coding standard

See #3 and https://github.com/nus-cs2103-AY2223S1/ip/pull/256

parnikkapore committed:
Lint

parnikkapore committed:
Documentation update

parnikkapore committed:
Add user guide

parnikkapore committed:
Update JavaDoc and error strings for unpacking

it's not called hydrating anymore!

parnikkapore committed:
Fix regression test

parnikkapore committed:
Update JavaFX to 18

The app does not run on M1 Macs as JavaFX 11 does not support them.
Let's update JavaFX to version 18 so the Apple peeps can enjoy the app
too.
